File: golang/src/p4/main.go
package main

import (
	"fmt"
	"strconv"
)

// largest palindrome of product of 2 3-digit numbers.
// Note that order doesn't matter (101x102 == 102x101)
func main() {
	palindromes := make([]int, 0)
	for x := 100; x <= 999; x++ {
		for y := 100; y <= 999; y++ {
			product := x * y
			if isPalindrome(product) {
				palindromes = append(palindromes, product)
			}
		}
	}

	max := 0
	for _, p := range palindromes {
		if p > max {
			max = p
		}
	}

	fmt.Println(max)
}

func isPalindrome(n int) bool {
	str := strconv.Itoa(n)
	stop := len(str) / 2
	for i, _ := range str {
		if i > stop {
			break
		}
		if str[i] != str[len(str)-i-1] {
			return false
		}
	}
	return true
}

File: cpp/src/p4/main.cc
#include <iostream>
#include <vector>
#include <stdlib.h>

bool isPalindrome(int n)
{
  const std::string s = std::to_string(n);
  const size_t stop = s.size() >> 1;
  for (size_t i = 0; i < stop; ++i) {
    if (s[i] != s[s.size()-i-1]) {
      return false;
    }
  }
  return true;
}

int main() {
  std::vector<int> palindromes;
  for (int x = 100; x <= 999; ++x) {
    for (int y = 100; y <= 999; ++y) {
      const int product = x*y;
      if (isPalindrome(product)) {
        palindromes.push_back(product);
      }
    }
  }

  int max = 0;
  for (const int x : palindromes) {
    max = (x > max ? x : max);
  }

  std::cout << max << "\n";
}
